`Development`: Fix refresh bug

marc-fett opened the pull request:
### Checklist

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://confluence.aet.cit.tum.de/spaces/AP/pages/257786006/Language+Guidelines+Client).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311714/PR+Guidelines).

#### Server

- [x] **Important**: I implemented the changes with a [very good performance](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311718/Performance+Guidelines) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311877/Server+Guidelines).
- [x] I documented the Java code using JavaDoc style.

#### Client

- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311716/Client+Guidelines).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311925/Client+Test+Guidelines).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Users were logged out after a short period when tabs were inactive or the browser throttled timers.
The goal is to stabilise session handling and make logout behaviour idempotent and user-friendly.

### Description

Most of the changes are refactoring changes to have a better overview.

**Server**

- /api/auth/logout endpoint is now permitAll and best-effort:
   - Clears cookies even when the refresh token is missing or already expired.
   - No longer throws UnauthorizedException if no refresh token is present.
- KeycloakAuthenticationService.invalidateRefreshToken() gracefully skips Keycloak logout when no refresh token is available.

**Client**

- Added descriptions for the major authentication classes
- Major refactor of the authentication layer:
   - Introduced AuthFacadeService to unify server-side (email/OTP) and Keycloak authentication flows.
   - Added new ServerAuthenticationService and KeycloakAuthenticationService with automatic token refresh and window listeners (focus, visibilitychange, online) to handle background-tab timer throttling.
   - Removed obsolete AuthenticationService, keycloak.service.ts, and related legacy classes.
   - Updated all components (login, registration, header, OTP input, etc.) to use the new facade and services.
   - Adjusted registration flow: replaced “verify” step with “otp”.
   - Added helper methods for updating the user profile and password.

### Steps for Testing

Prerequisites:
1. Deploy the branch to the test environment.
2. Clear all cookies and cache for the tumapply test website to start fresh.

**Hint: Copy the JWT to [jwt.io](https://www.jwt.io/) to check the expiration. Currently, the duration of the access token on the test server is 45 minutes. This will be changed once this PR is merged. If you want to test with a lower duration, feel free to ping me.**

Test cases:
1. Server logout without tokens
Log in with email & password, remove access_token & refresh token (via dev tools > Application), and click on logout.
✅ Response should be 200 OK and all auth cookies cleared. Redirect to the landing page. No user authenticated anymore.

2. Server login/refresh stability
Login with email/OTP and go to another tab. Return to the application and see that a refresh is triggered.
✅ User remains logged in; token refresh occurs on tab focus/visibility change.

3. Keycloak SSO flow
Same as 2. but with any Keycloak based login. For Keycloak-based login, you need to wait till the the token is expired. The token will only refresh if the expiration date is less than 20 seconds in the future. 
✅ Token refreshes automatically and session stays active.

4. Logout flows
Trigger logout from the UI for both server and Keycloak accounts.
✅ All cookies cleared, user redirected to start page, no error when tokens already expired.

5. Refresh tokens
Log in once with email/password and once with a Keycloak login (IdPs or TUM). Wait till the access token expires (for Keycloak it is in the headers of every call and for server it is in the cookies). 
✅ Around 20 seconds before the token expires, a refresh is triggered.

6. Please test all registration and login variations (otp, password, Google, TUM). And visit one secured URL (e.g., My Applications).
 
7. Any other authentication scenario you can imagine. :)

### Review Progress

#### Code Review

- [ ] Code Review 1

#### Manual Tests

- [x] Test 1

Bofan-Zhu commented:
When logging out, either normally or after clearing the tokens I'm being redirected to the applicant landing page (seems to be connected to the issue with that I'm redirected to /404 after login since it only happens when I stay on that page)

Bofan-Zhu commented:
For some reason when I register I am logged in after the second step and get an internal server error when I try to proceed to the fourth step.
<img width="445" height="157" alt="Screenshot 2025-09-19 at 13 20 53" src="https://github.com/user-attachments/assets/8b4af125-7304-4986-a0bb-64af0ee7d5b6" />
When I try to login via otp I can see that the code is sent out to my email but the modal closes unexpectedly after I click on "Continue" after entering my email

marc-fett commented:
> For some reason when I register I am logged in after the second step and get an internal server error when I try to proceed to the fourth step. <img alt="Screenshot 2025-09-19 at 13 20 53" width="445" height="157" src="https://private-user-images.githubusercontent.com/56017927/491548236-8b4af125-7304-4986-a0bb-64af0ee7d5b6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgzMTUxNjEsIm5iZiI6MTc1ODMxNDg2MSwicGF0aCI6Ii81NjAxNzkyNy80OTE1NDgyMzYtOGI0YWYxMjUtNzMwNC00OTg2LWEwYmItNjRhZjBlZTdkNWI2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTE5VDIwNDc0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJmYWRiNzFkZTIxZTJiYTFkZjc2NDM5OWYwZWYxNzllZmI0YjU2MjlhZDhmNDc4MDZlYWU5YTU3NmFjYTk1Y2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.X1kKOD_mt51u_yOf1Vjon4GEVPdx41YQEBsQubqVYQQ"> When I try to login via otp I can see that the code is sent out to my email but the modal closes unexpectedly after I click on "Continue" after entering my email

@Bofan-Zhu, thanks for your feedback. I fixed both issues now. 

The user gets redirected to the landing page when logging out and the current route is protected. Otherwise they stay on the route. 
I also fixed the issue when logging in with otp.

Bofan-Zhu commented:
Tested all listed scenarios, but the scenario regarding keycloak refresh does not work for me on the test server. When I change the focus to another tab, I can't see a refresh happening

ani-ib commented:
I tried registering using Apple and I was prompted to verify my email (see below screenshot) but I did not receive any email <img width="628" height="447" alt="image" src="https://github.com/user-attachments/assets/169e5944-0698-4fd4-8f29-b135f37255ae" />

marc-fett commented:
> I tried registering using Apple and I was prompted to verify my email (see below screenshot) but I did not receive any email <img alt="image" width="628" height="447" src="https://private-user-images.githubusercontent.com/88331280/491904649-169e5944-0698-4fd4-8f29-b135f37255ae.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgzNzg2OTgsIm5iZiI6MTc1ODM3ODM5OCwicGF0aCI6Ii84ODMzMTI4MC80OTE5MDQ2NDktMTY5ZTU5NDQtMDY5OC00ZmQ0LThmMjktYjEzNWYzNzI1NWFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTIwVDE0MjYzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ1NTViMDRiNmE4ZWI5NDRiYjNjYWI4NTRlM2MwZGEyZTE3YmVkMDc3ZDMyOTFlODE3ZjE0Mjg1MGYzMDA0M2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vCuF2cWK0NfTg_J3E81TNcz-O-MMsyFcWRVTWMHT82A">

Thank you for testing and reporting this. I fixed your issue, but there is another one, so I created an admin ticket.

marc-fett commented:
> > Tested all listed scenarios, but the scenario regarding keycloak refresh does not work for me on the test server. When I change the focus to another tab, I can't see a refresh happening. Login/Registration via Google also doesn't work for me: <img alt="Screenshot 2025-09-20 at 12 14 43" width="821" height="828" src="https://private-user-images.githubusercontent.com/56017927/491903102-d45ff5aa-1e92-45cc-93ae-35b653b165c4.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgzNjM2MzUsIm5iZiI6MTc1ODM2MzMzNSwicGF0aCI6Ii81NjAxNzkyNy80OTE5MDMxMDItZDQ1ZmY1YWEtMWU5Mi00NWNjLTkzYWUtMzViNjUzYjE2NWM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTIwVDEwMTUzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRmM2I1YTllZjA3NTI4YTg1MTNhYTQwNGRjZjY1MTE5NTY3ZDdiOGE5NDZjODEwZjQ5NjkwM2Y5OWJkYTMxMjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._2kYMWYTSoVt3OcWlnaSZZ0acJt5Tf6YquZOD6kw8t4">
> 
> I also tested all scenarios and don't see a refresh when logging in via keycloak when switching to another tab, but login via Google works for me (although I don't see a refresh when switching tabs via Google either)

Sorry for the confusion. When logging in with any Keycloak-based login method (either TUM login or IdPs), the application is checking the state of the access token, but only refreshes when the token is expiring within the next 20 seconds.

az108 commented on src/main/webapp/app/core/auth/keycloak-authentication.service.ts:
```diff
@@ -0,0 +1,232 @@
+      });
+      this.startTokenRefreshScheduler();
+    } catch (err) {
+      console.error(`Login with provider ${provider} failed:`, err);
```
should this be a console.error or is it relevant for the user, if yes please use a toast or similar

az108 commented on src/main/webapp/app/core/auth/server-authentication.service.ts:
```diff
@@ -0,0 +1,205 @@
+ */
+@Injectable({ providedIn: 'root' })
+export class ServerAuthenticationService {
+  private refreshTimerId?: number;
```
why are you not using signals here?

marc-fett replied:
The value is only used internally. No UI or other classes are using this variable. Therefore, a signal would be unnecessary overhead.

az108 commented on src/main/webapp/app/shared/components/organisms/registration/registration.ts:
```diff
@@ -56,44 +53,48 @@ export class Registration {
+    await this.runWithOrchestratorBusy(async () => {
+      await this.accountService.updateUser(firstName, lastName);
+      this.authOrchestrator.nextStep();
+    }, 'Could not update your profile name. Please try again.');
```
will this be visible to the user and if yes, do we need a german version? and will it be clear visible ?

az108 commented on src/main/webapp/app/shared/components/organisms/registration/registration.ts:
```diff
@@ -56,44 +53,48 @@ export class Registration {
+    await this.runWithOrchestratorBusy(async () => {
+      await this.accountService.updatePassword(password);
+      this.authOrchestrator.nextStep();
+    }, 'Could not update your password. Please try again.');
```
same here

marc-fett commented on src/main/webapp/app/core/auth/keycloak-authentication.service.ts:
```diff
@@ -0,0 +1,232 @@
+  });
+  private refreshIntervalId: ReturnType<typeof setInterval> | undefined;
+  private refreshInFlight: Promise<void> | null = null;
+  private windowListenersActive = false;
```
The value is only used internally. No UI or other classes are using this variable. Therefore, a signal would be unnecessary overhead.
